src/main/scala/zio/actors/ActorSystem.scala

```diff
@@ -0,0 +1,312 @@
+   * @tparam S ...
+   * @tparam E ...
+   * @tparam F ...
+   * @return
+   */
+  def createActor[S, E <: Throwable, F[+_]](
+    name: String,
+    sup: Supervisor[E],
+    init: S,
+    stateful: Stateful[S, E, F]
+  ): UIO[ActorRef[E, F]]
+
+  /**
+   *
+   * @param path ...
+   * @tparam E ...
+   * @tparam F ...
+   * @return
+   */
+  def selectActor[E <: Throwable, F[+_]](path: String): Task[ActorRef[E, F]]
+
+}
+
+/* INTERNAL API */
+
+private[actors] object ActorSystemUtils {
+
+  def resolvePath(path: String): Task[(String, Addr, Port, String)] = {
+
+    val regex = "^(?:zio:\\/\\/)(\\w+)[@](\\d+\\.\\d+\\.\\d+\\.\\d+)[:](\\d+)[/]([\\w|\\/]+)$".r
+
+    regex.findFirstMatchIn(path) match {
+      case Some(value) if value.groupCount == 4 =>
+        val actorSystemName = value.group(1)
+        val address         = value.group(2)
+        val port            = value.group(3).toInt
+        val actorName       = "/" + value.group(4)
+        IO.succeed((actorSystemName, address, port, actorName))
+      case None =>
+        IO.fail(
+          new Exception(
+            "Invalid path provided. The pattern is zio://YOUR_ACTOR_SYSTEM_NAME@ADDRES:PORT/RELATIVE_ACTOR_PATH"
+          )
+        )
+    }
+
+  }
+
+  def buildPath(actorSystemName: String, actorPath: String, remoteConfig: Option[(Addr, Port)]): String =
+    s"zio://$actorSystemName@${remoteConfig.map({ case (addr, port) => addr + ":" + port }).getOrElse("local")}$actorPath"
+
+  def readFromWire(socket: AsynchronousSocketChannel): Task[Any] =
+    for {
+      size      <- socket.read(4)
+      buffer    <- Buffer.byte(size)
+      intBuffer <- buffer.asIntBuffer
+      toRead    <- intBuffer.get(0)
+      content   <- socket.read(toRead)
+      arr       = content.toArray
+      obj       = new ObjectInputStream(new ByteArrayInputStream(arr)).readObject()
+    } yield obj
+
+  def writeToWire(socket: AsynchronousSocketChannel, obj: Any): Task[Unit] =
+    for {
+      stream <- IO.effect(new ByteArrayOutputStream())
+      oos    <- UIO.effectTotal(new ObjectOutputStream(stream))
+      _      = oos.writeObject(obj)
+      _      = oos.close()
+      bytes  = stream.toByteArray
+      _      <- socket.write(Chunk.fromArray(ByteBuffer.allocate(4).putInt(bytes.size).array()))
+      _      <- socket.write(Chunk.fromArray(bytes))
+    } yield ()
+
+}
+
+private[actors] case class ContextImpl[E <: Throwable, F[+_]](path: String, actorSystem: ActorSystem)
+    extends Context[E, F] {
+
+  override def self: Task[ActorRef[E, F]] = actorSystem.selectActor(path)
+
+  override def createActor[S, E1 <: Throwable, F1[+_]](
+    name: String,
+    sup: Supervisor[E1],
+    init: S,
+    stateful: Stateful[S, E1, F1]
+  ): UIO[ActorRef[E1, F1]] =
+    actorSystem.createActor(name, sup, init, stateful)
+
+  override def selectActor[E1 <: Throwable, F1[+_]](path: String): Task[ActorRef[E1, F1]] =
+    actorSystem.selectActor(path)
+}
+
+private[actors] case class ActorSystemImpl(
```

Comment:
I'd consider dropping these *Impl classes.

Reply:
Do you mean renaming them?

Reply:
Yes, perhaps even "merging" the trait and the implementation.

Reply:
Now I get it. Sure, I removed traits and left only case classes.

Reply:
Hmm with case classes like this the case class constructors are accessible, where they shouldn't, even with `private` modifier: `final case class Context private (...)`.

So to make them not accessible to users should I drop `case` and make them plain classes, go back to traits? Or something else?

Reply:
Ok at this moment I think that `Context` and `ActorSystem` shouldn't be `case classes` because they shouldn't be Serializable, where case classes are.